File: hello-python/src/audio_server/memory/memory_mapped_cache.py
"""Memory-mapped cache for efficient file I/O operations"""

import mmap
import os
import threading
from typing import Optional, IO
from loguru import logger

DEFAULT_PAGE_SIZE = 64 * 1024 * 1024
MAX_CACHE_SIZE = 2 * 1024 * 1024 * 1024


class MemoryMappedCache:
    """
    Memory-mapped cache implementation for efficient file operations.
    Thread-safe with RLock for concurrent access.
    """

    def __init__(self, path: str):
        self.path = path
        self.file: Optional[IO[bytes]] = None
        self.mmap: Optional[mmap.mmap] = None
        self.size = 0
        self.is_open = False
        self._lock = threading.RLock()  # Reentrant lock for thread safety

    def create(self, path: str, initial_size: int = 0) -> bool:
        with self._lock:
            return self._create_internal(path, initial_size)

    def _create_internal(self, path: str, initial_size: int = 0) -> bool:
        logger.debug(
            f"Creating mmap file: {path} with initial size: {initial_size}")

        try:
            # Ensure parent directory exists
            parent_dir = os.path.dirname(path)
            if parent_dir:
                os.makedirs(parent_dir, exist_ok=True)

            if os.path.exists(path):
                os.remove(path)

            # Create new file with w+b mode (write + binary, creates if not exists)
            self.file = open(path, 'w+b')
            if initial_size > 0:
                self.file.truncate(initial_size)
                self.size = initial_size
            else:
                self.size = 0

            if self.size > 0:
                self.map_file()

            self.is_open = True
            logger.debug(
                f"Created mmap file: {path} with size: {initial_size}")
            return True

        except Exception as e:
            logger.error(f"Error creating file {path}: {e}")
            return False

    def open(self, path: str) -> bool:
        with self._lock:
            return self._open_internal(path)

    def _open_internal(self, path: str) -> bool:
        logger.debug(f"Opening mmap file: {path}")

        try:
            if not os.path.exists(path):
                logger.error(f"File does not exist: {path}")
                return False

            self.file = open(path, 'r+b')
            self.size = os.path.getsize(path)

            if self.size > 0:
                self.map_file()

            self.is_open = True
            logger.debug(f"Opened mmap file: {path} with size: {self.size}")
            return True

        except Exception as e:
            logger.error(f"Error opening file {path}: {e}")
            return False

    def close(self) -> None:
        with self._lock:
            if self.is_open:
                self.unmap_file()
                logger.debug(f"Closed mmap file: {self.path}")

    def write(self, offset: int, data: bytes) -> int:
        with self._lock:
            try:
                if not self.is_open or self.mmap is None:
                    initial_size = offset + len(data)
                    if not self._create_internal(self.path, initial_size):
                        return 0

                required_size = offset + len(data)
                if required_size > self.size:
                    if not self._resize_internal(required_size):
                        logger.error(
                            "Failed to resize file for write operation")
                        return 0

                assert self.mmap is not None
                self.mmap[offset:offset + len(data)] = data

                logger.debug(
                    f"Wrote {len(data)} bytes to {self.path} at offset {offset}")
                return len(data)

            except Exception as e:
                logger.error(f"Error writing to mapped file {self.path}: {e}")
                return 0

    def read(self, offset: int, length: int) -> bytes:
        with self._lock:
            try:
                if not self.is_open or self.mmap is None:
                    logger.debug(
                        f"File not open, attempting to open for reading: {self.path}")
                    if not self._open_internal(self.path):
                        logger.error(
                            f"Failed to open file for reading: {self.path}")
                        return b''

                if offset >= self.size:
                    logger.debug(
                        f"Read offset {offset} at or beyond file size {self.size} - end of file")
                    return b''

                actual_length = min(length, self.size - offset)
                assert self.mmap is not None
                data = self.mmap[offset:offset + actual_length]

                logger.debug(
                    f"Read {actual_length} bytes from {self.path} at offset {offset}")
                return data

            except Exception as e:
                logger.error(
                    f"Error reading from mapped file {self.path}: {e}")
                return b''

    def get_size(self) -> int:
        with self._lock:
            return self.size

    def get_path(self) -> str:
        return self.path

    def get_is_open(self) -> bool:
        with self._lock:
            return self.is_open

    def resize(self, new_size: int) -> bool:
        with self._lock:
            return self._resize_internal(new_size)

    def _resize_internal(self, new_size: int) -> bool:
        try:
            if not self.is_open:
                logger.error(f"File not open for resize: {self.path}")
                return False

            if new_size == self.size:
                return True

            # Unmap only, don't close the file
            if self.mmap is not None:
                self.mmap.close()
                self.mmap = None

            assert self.file is not None
            self.file.truncate(new_size)
            self.size = new_size

            if self.size > 0:
                self.map_file()

            logger.debug(
                f"Resized and remapped file {self.path} to {new_size} bytes")
            return True

        except Exception as e:
            logger.error(f"Error resizing file {self.path}: {e}")
            return False

    def finalize(self, final_size: int) -> bool:
        with self._lock:
            try:
                if not self.is_open:
                    logger.warning(
                        f"File not open for finalization: {self.path}")
                    return False

                if not self._resize_internal(final_size):
                    logger.error(
                        f"Failed to resize file during finalization: {self.path}")
                    return False

                if self.mmap is not None:
                    self.mmap.flush()

                logger.debug(
                    f"Finalized file: {self.path} with size: {final_size}")
                return True

            except Exception as e:
                logger.error(f"Error finalizing file {self.path}: {e}")
                return False

    def map_file(self) -> None:
        if self.file is None:
            raise ValueError("File is not open")

        self.mmap = mmap.mmap(self.file.fileno(), 0)
        logger.debug(
            f"Successfully mapped file: {self.path} ({self.size} bytes)")

    def unmap_file(self) -> None:
        if self.mmap is not None:
            self.mmap.close()
            self.mmap = None

        if self.file is not None:
            self.file.close()
            self.file = None

        self.is_open = False

    def __del__(self):
        self.close()


# Alias for backward compatibility
MmapCache = MemoryMappedCache

File: hello-kotlin/src/main/kotlin/server/memory/MemoryMappedCache.kt
// Memory-mapped cache for efficient file I/O.
// Provides write, read, resize, and finalize operations.
// Matches Python MmapCache functionality.

package server.memory

import server.Logger
import java.io.RandomAccessFile
import java.nio.MappedByteBuffer
import java.nio.channels.FileChannel
import java.nio.file.Files
import java.nio.file.Paths
import java.util.concurrent.locks.ReentrantReadWriteLock
import kotlin.concurrent.read
import kotlin.concurrent.write

const val DEFAULT_PAGE_SIZE = 64L * 1024 * 1024 // 64MB
const val MAX_CACHE_SIZE = 2L * 1024 * 1024 * 1024 // 2GB

/**
 * Memory-mapped cache implementation using Java's MappedByteBuffer.
 * Thread-safe with ReadWriteLock for concurrent access.
 */
class MemoryMappedCache(private val path: String) {
    private var file: RandomAccessFile? = null
    private var channel: FileChannel? = null
    private var mappedBuffer: MappedByteBuffer? = null
    private var size: Long = 0
    private var isOpen: Boolean = false
    private val rwLock = ReentrantReadWriteLock()

    /**
     * Create a new memory-mapped file.
     */
    fun create(initialSize: Long = 0): Boolean = rwLock.write {
        return try {
            // Remove existing file
            val filePath = Paths.get(path)
            if (Files.exists(filePath)) {
                Files.delete(filePath)
            }

            // Create and open file
            file = RandomAccessFile(path, "rw")
            channel = file!!.channel

            if (initialSize > 0) {
                // Set file size
                file!!.setLength(initialSize)
                size = initialSize

                // Map file into memory
                mapFile()
            } else {
                size = 0
            }

            isOpen = true
            Logger.debug("Created mmap file: $path with size: $initialSize")
            true
        } catch (e: Exception) {
            Logger.error("Error creating file $path: ${e.message}")
            close()
            false
        }
    }

    /**
     * Open an existing memory-mapped file.
     */
    fun open(): Boolean = rwLock.write {
        try {
            val filePath = Paths.get(path)
            if (!Files.exists(filePath)) {
                Logger.error("File does not exist: $path")
                return@write false
            }

            file = RandomAccessFile(path, "rw")
            channel = file!!.channel
            size = java.io.File(path).length()

            if (size > 0) {
                mapFile()
            }

            isOpen = true
            Logger.debug("Opened mmap file: $path with size: $size")
            true
        } catch (e: Exception) {
            Logger.error("Error opening file $path: ${e.message}")
            closeInternal()
            false
        }
    }

    /**
     * Close memory-mapped file.
     */
    fun close() = rwLock.write {
        closeInternal()
    }

    /**
     * Internal close without lock (called from within locked methods).
     */
    private fun closeInternal() {
        if (isOpen) {
            try {
                unmapFile()
            } catch (e: Exception) {
                Logger.error("Error closing file $path: ${e.message}")
            } finally {
                isOpen = false
            }
        }
    }

    /**
     * Write data to memory-mapped file.
     */
    fun write(offset: Long, data: ByteArray): Int = rwLock.write {
        try {
            // If file is not open, create it
            if (!isOpen) {
                val initialSize = offset + data.size
                if (!createInternal(initialSize)) {
                    return@write 0
                }
            }

            val requiredSize = offset + data.size
            
            // If file needs to grow or has no mapped buffer, resize it
            if (requiredSize > size || mappedBuffer == null) {
                val newSize = maxOf(requiredSize, size)
                if (!resizeInternal(newSize)) {
                    Logger.error("Failed to resize file for write operation")
                    return@write 0
                }
            }

            // Write to memory-mapped buffer
            mappedBuffer!!.position(offset.toInt())
            mappedBuffer!!.put(data)
            data.size
        } catch (e: Exception) {
            Logger.error("Error writing to file $path: ${e.message}")
            0
        }
    }

    /**
     * Read data from memory-mapped file.
     */
    fun read(offset: Long, length: Int): ByteArray = rwLock.write {
        try {
            if (!isOpen || mappedBuffer == null) {
                if (!openInternal()) {
                    Logger.error("Failed to open file for reading: $path")
                    return@write byteArrayOf()
                }
            }

            if (offset >= size) {
                return@write byteArrayOf()
            }

            // Read from memory-mapped buffer
            val actualLength = length.coerceAtMost((size - offset).toInt())
            val data = ByteArray(actualLength)

            mappedBuffer!!.position(offset.toInt())
            mappedBuffer!!.get(data)

            Logger.debug("Read $actualLength bytes from $path at offset $offset")
            data
        } catch (e: Exception) {
            Logger.error("Error reading from file $path: ${e.message}")
            byteArrayOf()
        }
    }

    /**
     * Get the size of the file.
     */
    fun getSize(): Long = rwLock.read { size }

    /**
     * Check if the file is open.
     */
    fun getIsOpen(): Boolean = rwLock.read { isOpen }

    /**
     * Resize the file to a new size (internal, no lock).
     */
    private fun resizeInternal(newSize: Long): Boolean {
        return try {
            if (!isOpen) {
                Logger.error("File not open for resize: $path")
                return false
            }

            if (newSize == size) {
                return true
            }

            // Unmap current buffer (but don't close file/channel)
            mappedBuffer = null

            // Resize file
            if (newSize < size) {
                Logger.warning("Truncating file $path to $newSize")
            }

            file!!.setLength(newSize)
            size = newSize

            // Remap file
            if (size > 0) {
                mapFile()
            }

            Logger.debug("Resized and remapped file $path to $newSize bytes")
            true
        } catch (e: Exception) {
            Logger.error("Error resizing file $path: ${e.message}")
            e.printStackTrace()
            false
        }
    }

    /**
     * Finalize the file to its final size.
     */
    fun finalize(finalSize: Long): Boolean = rwLock.write {
        try {
            if (!isOpen) {
                Logger.warning("File not open for finalization: $path")
                return@write false
            }

            if (!resizeInternal(finalSize)) {
                Logger.error("Failed to resize file during finalization: $path")
                return@write false
            }

            // Force mapped buffer to write to disk
            mappedBuffer?.force()

            Logger.debug("Finalized file: $path with size: $finalSize")
            true
        } catch (e: Exception) {
            Logger.error("Error finalizing file $path: ${e.message}")
            false
        }
    }

    /**
     * Create internal without lock.
     */
    private fun createInternal(initialSize: Long): Boolean {
        return try {
            // Remove existing file
            val filePath = Paths.get(path)
            if (Files.exists(filePath)) {
                Files.delete(filePath)
            }

            // Create and open file
            file = RandomAccessFile(path, "rw")
            channel = file!!.channel

            if (initialSize > 0) {
                // Set file size
                file!!.setLength(initialSize)
                size = initialSize

                // Map file into memory
                mapFile()
            } else {
                size = 0
            }

            isOpen = true
            Logger.debug("Created mmap file: $path with size: $initialSize")
            true
        } catch (e: Exception) {
            Logger.error("Error creating file $path: ${e.message}")
            closeInternal()
            false
        }
    }

    /**
     * Open internal without lock.
     */
    private fun openInternal(): Boolean {
        return try {
            val filePath = Paths.get(path)
            if (!Files.exists(filePath)) {
                Logger.error("File does not exist: $path")
                return false
            }

            file = RandomAccessFile(path, "rw")
            channel = file!!.channel
            size = java.io.File(path).length()

            if (size > 0) {
                mapFile()
            }

            isOpen = true
            Logger.debug("Opened mmap file: $path with size: $size")
            true
        } catch (e: Exception) {
            Logger.error("Error opening file $path: ${e.message}")
            closeInternal()
            false
        }
    }

    /**
     * Map the file into memory using MappedByteBuffer.
     */
    private fun mapFile() {
        if (channel != null && size > 0) {
            // Map entire file into memory (READ_WRITE mode for both reading and writing)
            mappedBuffer = channel!!.map(
                FileChannel.MapMode.READ_WRITE,
                0,
                size
            )
            Logger.debug("Successfully mapped file: $path ($size bytes)")
        }
    }

    /**
     * Unmap the file from memory and close resources.
     */
    private fun unmapFile() {
        mappedBuffer = null

        if (channel != null) {
            try {
                channel!!.close()
            } catch (e: Exception) {
                Logger.error("Error closing channel: ${e.message}")
            }
            channel = null
        }

        if (file != null) {
            try {
                file!!.close()
            } catch (e: Exception) {
                Logger.error("Error closing file: ${e.message}")
            }
            file = null
        }
    }
}
